numpy/core/src/umath/ufunc_object.c

```diff
@@ -3707,24 +3707,22 @@ PyUFunc_GenericReduction(PyUFuncObject *ufunc, PyObject *args,
             PyDict_SetItem(kwds, npy_um_str_out, out_obj);
         }
     }
-            
+
     if (operation == UFUNC_REDUCEAT) {
         PyArray_Descr *indtype;
         indtype = PyArray_DescrFromType(NPY_INTP);
         if (!PyArg_ParseTupleAndKeywords(args, kwds, "OO|OO&O&:reduceat", reduceat_kwlist,
-                                        &op,
-                                        &obj_ind,
-                                        &axes_in,
-                                        PyArray_DescrConverter2, &otype,
-                                        PyArray_OutputConverter, &out)) {
-            Py_XDECREF(otype);
```

Comment:
`otype` cannot be defined if this fails.

Reply:
That's not true - this was correct before - it could fail on the `OutputConverter`.

Reply:
In fact, this should maybe xdecref `out` too.

Reply:
I looked for other usages, and they always DECREF only stuff before `OutputConverter` - so, I guess one does not have to worry about the one that does fail, just what comes before. Anyway, I'll change it back.

Reply:
Or, rather, change it to `goto fail` as it is for the other options.

Reply:
> I guess one does not have to worry about the one that does fail, just what comes before.

Looking at the implementation of `PyArg_ParseTupleAndKeywords` in 3.6, I think I can break this by providing an argument by keyword and position, which happens after the other arguments are parsed. I certainly don't think that cpython guarantees the last object will never need freeing

Reply:
> I guess one does not have to worry about the one that does fail, just what comes before.

Looking at the implementation of `PyArg_ParseTupleAndKeywords` in 3.6, I think I can break this by providing an argument by keyword and position, which happens after the other arguments are parsed. I certainly don't think that cpython guarantees the last object will never need freeing

Reply:
You may well be right, but since this is separate from the fix here (in that it is done *everywhere* `OutputConverter` is used), I think is should be a separate issue: #10197

Reply:
I'm wrong, `PyArray_OutputConverter` borrows a reference.